### ISSUE TITLE
Using a standalone service class instance in a vue instance

### ISSUE THREAD

**Issue by JosephSilber** [vuejs/vue#1787]
I'm used to encapsulating functionality in services/models in server-side code as well as in Angular.

These services/models have their own data/state and their own methods.

But in Vue, data and methods are completely separated; `data` for state and `methods` for functions. So it seems there's no place to set an instance of a model/service.

The closest thing I can think of is to use mixins, but they just go into the flat list of data and methods, which is extremely inelegant and will inevitably lead to conflicts (this would be _a little_ better if we were able to have [nested methods](https://github.com/vuejs/vue/issues/1470)).

The only other option I can think of would be to set it directly on the vm in `ready`/`compiled`, but then its data won't be reactive, which kind of defeats the purpose.

Is anything like this possible, or is there another way to accomplish it?

**Comment by nadir-isen** [comment]
If you want to do Angular with Vue, why don't you use Angular directly?
;-)
Vue is Angular without the heaviness

**Comment by boussou** [comment]
If you want to do Angular with Vue, why don't you use Angular directly?
;-)
Vue is Angular without the heaviness

**Comment by DarrylD** [comment]
Looking for something like this also as Angular isn't an option in some cases.

**Comment by boussou** [comment]
I would add: Having data **gathered** in one place is a good thing for many reasons.
(and I really don't understand the sentence "data and methods are completely separated", it is actually the opposite IMHO)

**Comment by JosephSilber** [comment]
@boussou I just used Angular as an example. This has nothing to do with Angular and is just a good practice all around.

I want to have small classes that encapsulated their own functionality, without being bound to vue at all.

Vue should be able to utilize those.

**Comment by boussou** [comment]
got it.

**Comment by yyx990803** [comment]
This is probably where Vue does get "opnionated": reactive state, i.e. the source of truth for your UI, needs to be plain objects.

However, this doesn't prevent you from sharing these reactive objects between a Vue instance and a class of your choice. Rather than thinking that your service class instance should "own" the data, think it more like both the Vue instance and your service class instance are "sharing" the data.

So something like this could probably work:

``` js
new Vue({
  created () {
    this.myService = new MyService(this.someDataTheServiceShouldManage)
  }
})
```

**Comment by JosephSilber** [comment]
@yyx990803 Sorry if I sound ungrateful, but that sounds lame to me.

1. The service should probably be in charge of fetching its own data (can maybe be remedied by having a static function on the service for that, then calling it from `route.data`).
2. This wouldn't allow the view to call any of the service's methods. It'd mean you have to create proxy methods for every service method you may want to allow the view to access. That's just sad.

---

Here's an example:

```js
import Order from 'order';

Vue.component('order-page', {

    route: {
        data(transition) {
            Order.fetch(transition.to.params.id).then(order => {
                return { orderData: order }
            })
        }
    },

    data() {
        return {
            orderData: {}
        }
    },

    created() {
        this.order = new Order(this.orderData)
    },

    methods: {
        saveOrder() {
            this.order.save()
        },
        deleteOrder() {
            this.order.delete()
        }
        .......
    }
})
```

And I'd then have to bind in the view for example to `orderData.total` instead of `order.total`.

This looks quite lame to me :disappointed:

**Comment by yyx990803** [comment]
You can just do `@click="order.save()"`, just remember to include the parens, no need to proxy them with methods...

**Comment by yyx990803** [comment]
I'm closing this since this is more like an open-ended discussion - feel free to continue this on the forums.

**Comment by phil294** [comment]
@JosephSilber you can now do this using Composition API (included in upcoming 3.0 by default)
https://vue-composition-api-rfc.netlify.com/
